src/core/shape_inference/include/topk_shape_inference.hpp

```diff
@@ -13,12 +13,44 @@ namespace ov {
 namespace op {
 namespace v1 {
 
-template <typename T>
-void shape_infer(const TopK* op,
-                 const std::vector<T>& input_shapes,
-                 std::vector<T>& output_shapes,
-                 const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data = {}) {
-    NODE_VALIDATION_CHECK(op, (input_shapes.size() == 2 && output_shapes.size() == 2));
+// Helper to get correct K from tensor as shape.
+template <class T>
+struct GetK {
```

Comment:
I'd rather see this as a free function or a lambda. Is there any reason to wrap this in a struct?

Reply:
Yes, to pass it as functor in `get_input_const_data_as_shape` and correct deduce type in operator() for input which got fundamental type of tensor data element.
Free function used here as template need to be specified and one "result type" is know as `TDimValue` but 2nd is runtime element type in tensor.
Lambda not support `auto` and if I use e.g. `TDimValue` it will be `size_t` or `int64_t`, then k value validation can give false results.

Assume `k` tensor is `i64` and we do inference for static shape so dimension value type is `size_t`
then for values -1,-2, etc. the tensor value will be casted to `size_t` which give positive number and always pass the check
or when `size_t` is 32bit then values greater than (int64_max/2) will be casted to 0..u32.max instead be rejected as too large.